### PR TITLE
Block editor: remove `__unstableElementContext` and filter `EditorStyles` instead

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

TO DO: do think for every filter using `BlockList.__unstableElementContext`, but if it works for Duotone, it will work for the rest too. We can look at these in separate PRs.

This PR aims to eliminate `BlockList.__unstableElementContext`, which is currently only necessary to add style elements and SVGs. While working on [Allow styles to be changed dynamically through editor settings](https://github.com/WordPress/gutenberg/pull/52767), I realised that we could use the `styles` block editor setting instead. I also realised that we're actually using this for the [Duotone Global Styles Presets](https://github.com/WordPress/gutenberg/blob/dc27d6b034a0d112dce0b71fb7d3d5dd8e870fe6/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L1182), so this make it even more appealing.

What we can do is filter the editor styles setting through the `EditorStyles` component through context: pass down a callback to update/clear a given style with id. Not only can this be used to add new styles with new ID, but you can also filter/take over server rendered styles.

For now, this is better because we no longer need to render anything in the React tree and it makes it easier to design a new API.

## Why?

Currently, we can't refactor `editor.BlockListBlock` because of the need to portal styles and svgs. This PR removes this need.

### CODE REVIEW

File: packages/block-editor/src/components/global-styles/use-global-styles-output.js [hunk 1]
```diff
@@ -163,11 +163,9 @@ function getPresetsSvgFilters( blockPresets = {} ) {
 			.filter( ( origin ) => presetByOrigin[ origin ] )
 			.flatMap( ( origin ) =>
 				presetByOrigin[ origin ].map( ( preset ) =>
-					renderToString(
```
Comment (ellatrix) [inline packages/block-editor/src/components/global-styles/use-global-styles-output.js:167]: Notice how we also get rid of this `renderToString` call.

File: packages/block-editor/src/hooks/duotone.js [hunk 2]
```diff
@@ -290,25 +282,46 @@ function DuotoneStyles( {
...
+			id: `duotone-${ filterId }`,
+			assets:
+				colors !== 'unset' ? getDuotoneFilter( filterId, colors ) : '',
+			__unstableType: 'svgs',
+		} );
+
+		updateSettings( {
+			...settings,
+			styles,
+		} );
```
Comment (ellatrix) [inline packages/block-editor/src/hooks/duotone.js]: All the above is very awkward, we'll need to figure out a better way to do it when we design a block support API.

**Comment by Mamaduka** [comment]
@ellatrix, the description mentions  `Allow styles to be changed dynamically through editor settings.` but it links to the same branch. Is there a different branch/PR for that API?

**Comment by ellatrix** [comment]
@Mamaduka Fixed the link. Github has changed some of the links to relative hash links, so when I paste links into Github they assume its the same PR...

**Comment by Mamaduka** [comment]
I like the idea of keeping every style preset in a single location/setting. This matches behavior on the PHP side.

## Concerns
* The `useEffect` are sometimes easy to mess up, and it might cause settings updates on every render. How can we ensure stability if this becomes part of the public API?
* Race conditions could happen when updating editor settings.

**Comment by ellatrix** [comment]
A proposal to prevent the race conditions: #53516

**Comment by ellatrix** [comment]
Ok, rewrote the PR to filter the styles instead of going through editor settings.

File: packages/block-editor/src/components/editor-styles/index.js [hunk 3]
```diff
@@ -67,11 +74,42 @@ function useDarkThemeBodyClassName( styles ) {
...
+			const index = _styles.findIndex( ( { id } ) => id === override.id );
+			if ( index === -1 ) {
+				_styles.push( override );
+			} else {
+				_styles[ index ] = override;
+			}
+		}
+
+		return _styles;
+	}, [ styles, overrides ] );
```
Comment (Mamaduka) [inline packages/block-editor/src/components/editor-styles/index.js]: Super early feedback: I think all these style/svg transformations and iterations could be absorbed in a single `useMemo`.

File: packages/block-editor/src/components/editor-styles/index.js [hunk 4]
```diff
@@ -19,6 +24,8 @@ import transformStyles from '../../utils/transform-styles';
 const EDITOR_STYLES_SELECTOR = '.editor-styles-wrapper';
 extend( [ namesPlugin, a11yPlugin ] );
 
+export const updateStyleContext = createContext();
```
Comment (Mamaduka) [inline packages/block-editor/src/components/editor-styles/index.js]: Should we provide a `defaultValue` here, something like a `noop`?

**Review by andrewserong — COMMENTED** [review]
> What we can do is filter the editor styles setting through the EditorStyles component through context: pass down a callback to update/clear a given style with id. 

This looks very cool! I love how this avoids having to use a portal, and that it makes for a much clearer approach to injecting styles, particularly with providing an `id`.

Do you think this approach would also work for if / when we need to use the same styles output in a non-editor context? That is, if we were to try to implement a JS-based dynamic rendering (as in #54047) for rendering out final HTML that includes all generated styles from block supports, could we re-use the same logic?

I suppose in concrete terms, I'm wondering if it'd be valuable for the `updateStyleContext` provider to potentially be separate from `EditorStyles` so that it could be used by both `EditorStyles` and a separate renderer of some kind.

Apologies if all that's a bit off topic from the task at hand, I've only just started to think about some of the things in #54047 and at least at first glance it looks like this PR could be very useful for it / potentially get a long way there! 🙂

**Comment by ellatrix** [comment]
@andrewserong I'm not entirely following what you want to achieve? Could you dumb it down for me? 😅 I mean, sure, we can separate it from EditorStyles in a follow up when a use case appears. My number 1 priority with this PR is removing the need for block supports API stuff to render elements in the React tree (`createPortal` requires rendering in the React tree). With this PR we don't need to touch the tree and everything can simply be hooks.

**Comment by andrewserong** [comment]
> @andrewserong I'm not entirely following what you want to achieve? Could you dumb it down for me? 😅 I mean, sure, we can separate it from EditorStyles in a follow up when a use case appears.

Sure, my apologies, I could have been clearer with my thinking here! 🙂 The use case I had in mind is the one that @youknowriad described over in #53874, and was then split out into #54047:

* A third party editor builds a custom editor using the block-editor package
* Within the editor, block supports styles work as expected due to editor styles being injected correctly
* The final serialized markup using this custom editor does not contain Duotone, Layout, and Element block supports styles, as they're intended to be dynamically rendered / injected
* Because this custom editor has no PHP counterpart, the third party editor needs a mechanism to generate and render the the above styles

One approach, discussed in #54047, is the idea of having a flag and/or API of some kind so that a 3rd party editor could retrieve the serialized markup with the generated styles all included. Essentially this would be a JS-based output of rendered block markup that pretty much matches the PHP-based server rendering that's currently being used.

Where it relates to this PR, is that if we have a mechanism that block supports can use to inject styles into EditorStyles, it'd be great if this could be an API that could be used universally for either editor styles or styles for rendered HTML intended for display, or to be saved to a database somewhere. So, what I was wondering is whether `updateStyle` as it's being called within `hooks` could be not tied to editor styles per se, but be a context that could be shared between editor styles and a future `RenderedStyles` or something like that?

> My number 1 priority with this PR is removing the need for block supports API stuff to render elements in the React tree (createPortal requires rendering in the React tree). With this PR we don't need to touch the tree and everything can simply be hooks.

That sounds like a really good goal to me! In terms of the stuff surrounding rendered styles, most of that likely doesn't need to be worried about for this PR. My only real question at this stage is whether it'd be worth splitting context out to be more generic, and overall if this is a direction that could help with the work that Riad's looking into for 3rd party editors. So far it looks like a promising direction to me!

@youknowriad did you have any thoughts on whether this is something we should be concerned with for now, or would it be a question for future refactorings to support 3rd party editors?

**Comment by youknowriad** [comment]
@andrewserong In my mind, it's not really a flag but more of a new function to get the final HTML. and for the styles that we inject to the "element context" today, we'd also include them in that new "render" function. So as long as these styles are written in a declarative way by the block support, we could reuse them within the potentially new "render" function.

**Comment by andrewserong** [comment]
Okay, great. Thanks for confirming @youknowriad! @ellatrix I think from the sounds of it there's nothing to worry about in this PR, then, and we can explore our options for the render behaviour in follow-ups 🙂

**Comment by ellatrix** [comment]
Cool thanks! It sounds good if other ideas are explored. My goal with this PR however is to remove the portals and consolidate around `EditorStyles`, so we can move forward in refactoring the block hooks. This PR doesn't expose a new API, so we can still take another approach later if necessary.

**Comment by ellatrix** [comment]
@Mamaduka I addressed your feedback :)

File: packages/block-editor/src/hooks/duotone.js [hunk 5]
```diff
@@ -290,25 +267,47 @@ function DuotoneStyles( {
...
+	}, [
+		isValidFilter,
+		colors,
+		selector,
+		filterId,
+		setStyleOverride,
+		deleteStyleOverride,
+	] );
+
+	return null;
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/duotone.js:310]: Minor: If it's not a public Api, this `DuotoneStyles` component could probably be a `useDuotoneStyles` hook since it renders nothing.

**Comment by youknowriad** [comment]
I didn't test but this is very cool 👍

File: packages/block-editor/src/hooks/duotone.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/hooks/duotone.js:310]: Well, that's exactly the whole point of this PR 👍

Comment (ellatrix) [reply packages/block-editor/src/hooks/duotone.js:310]: Actually, I do want to wait a bit with until we refactor duotone more. There's a `shouldRender` condition that would be removed. It would be good if we can prevent a bunch of hooks like useSetting from doing anything if it's not needed for that block. I'd rather wait when we look into refactoring this whole API so we don't impact performance.

File: packages/block-editor/src/components/editor-styles/index.js [hunk 6]
```diff
@@ -112,6 +124,7 @@ export default function EditorStyles( { styles } ) {
 				} }
 				dangerouslySetInnerHTML={ { __html: transformedSvgs } }
 			/>
+			{ children }
```
Comment (Mamaduka) [inline packages/block-editor/src/components/editor-styles/index.js]: Do we still need to render `children`? The component is no longer a `Provider` wrapper - 715d47f62220dbf2854e2814d10fdb40912425df.

**Review by Mamaduka — COMMENTED** [review]
@ellatrix, I see the following error when deleting the block with Duotone filters applied.

## Screenshot

CleanShot 2023-09-09 at 12 49 13

**Comment by ellatrix** [comment]
Fixed

File: packages/block-editor/src/components/editor-styles/index.js [hunk 7]
```diff
@@ -9,15 +9,20 @@ import a11yPlugin from 'colord/plugins/a11y';
...
 /**
  * Internal dependencies
  */
 import transformStyles from '../../utils/transform-styles';
+import { store as blockEditorStore } from '../../store';
+import { unlock } from '../../lock-unlock';
 
 extend( [ namesPlugin, a11yPlugin ] );
 
+export const updateStyleContext = createContext( () => {} );
```
Comment (Mamaduka) [inline packages/block-editor/src/components/editor-styles/index.js]: I missed this one. I think this context is no longer needed.

**Review by Mamaduka — APPROVED** [review]
The duotone changes test well for me with various blocks. I couldn't spot any regressions.

File: packages/block-editor/src/components/editor-styles/index.js (hunk 7 above)
Comment (ellatrix) [reply packages/block-editor/src/components/editor-styles/index.js]: Sorry, badly rebased.

**Comment by youknowriad** [comment]
👏 @ellatrix

**Comment by tellthemachines** [comment]
Hi! This PR seems to have introduced the regression in #54956. I'm not 100% sure why though, would be great if @ellatrix or anyone else could help debug this!

**Comment by Mamaduka** [comment]
I wonder if this is the same issue that @andrewserong discovered here: https://github.com/WordPress/gutenberg/pull/54493#pullrequestreview-1632213052.

**Comment by tellthemachines** [comment]
I don't _think_ so, at least not if it's related to id collision, because the blocks that aren't getting their styles output all have different ids.

**Comment by andrewserong** [comment]
Just added a comment on the linked issue (https://github.com/WordPress/gutenberg/issues/54956#issuecomment-1744272889), this sure is a weird one! I'm wondering if previews might have issues since they have settings passed down to their block editor providers from above, so updates to state within the previews might not be propagating up to the "real" block editor store?

**Comment by ska-dev-1** [comment]
@ellatrix Any ideas why [this](https://github.com/WordPress/gutenberg/pull/52767#issuecomment-1643025927) method of adding styles to editor settings no longer works for me in the Site Editor but still works in Post Editor (WP 6.3.2/6.4-RC2).
In Site Editor `wp.data.select('core/editor').getEditorSettings().styles` returns undefined, while in Post Editor it's an array with already some styles in there. Even if I set the styles to an array myself the styles are not loaded.

As for `setStyleOverride`, every time a style/override is added/removed/updated

https://github.com/WordPress/gutenberg/blob/a84537bde532016fcc84387de9672b6b37e9a7a8/packages/block-editor/src/components/editor-styles/index.js#L98

it re-transforms every style:

https://github.com/WordPress/gutenberg/blob/a84537bde532016fcc84387de9672b6b37e9a7a8/packages/block-editor/src/components/editor-styles/index.js#L89

For me this is performing very badly as I can have hundreds of styles.
